File: hs/solve.hs
import Data.Maybe
import Debug.Trace

type Board = [Int]
type Group = [Int]

count :: Int -> [Int] -> Int
count elem array = length (filter (==elem) array)

digits :: [Int]
digits = [1 .. 9]

legalGroup :: Group -> Bool
legalGroup group = (<2) (foldr1 max [count d group | d <- digits])

getGroups :: Board -> [Group]
getGroups board = rows ++ cols ++ boxs
    where
        rows = [ [board !! (9*y+x) | x <- [0..8]]   | y <- [0..8] ]
        cols = [ [board !! (9*x+y) | x <- [0..8]]   | y <- [0..8] ]
        boxs = [ [board !! x | x <- [0..80], 3*(x `div` 27)+(x `mod` 9) `div` 3 == y] | y <- [0..8] ]

legalBoard :: Board -> Bool
legalBoard board = all legalGroup (getGroups board)

getEmptyPosition :: Board -> Int
getEmptyPosition board = getFirstZero board 0 
    where
    getFirstZero :: [Int] -> Int -> Int
    getFirstZero [] n = -1
    getFirstZero (x:xs) n | x == 0 = n
                          | otherwise = (getFirstZero xs (n+1))

spliceElement :: [Int] -> Int -> Int -> [Int]
spliceElement arr index elem = take index arr ++ [elem] ++ drop (index+1) arr



solveBoard :: Board -> [Board]
-- solveBoard board | trace ("current board: " ++ show board) False = undefined
solveBoard board | (pos == -1) = [board]
                 | otherwise = concat (map solveBoard 
                        (filter legalBoard
                             [spliceElement board pos d | d <- digits]
                        )
                    )
                 where pos = getEmptyPosition board

getSolution :: Board -> Maybe Board
getSolution board = getSolutionHelper (solveBoard board)
    where
        getSolutionHelper [] = Nothing
        getSolutionHelper (x:xs) = Just x

reprBoard :: Board -> String
reprBoard board = foldr1 (++) [(foldl (++) "" (map show r)) ++ "\n" | r <- rows]
    where rows = [ [board !! (9*y+x) | x <- [0..8]] | y <- [0..8] ]


main = do
    fileContents <- readFile "../sudoku_puzzles"
    let boards = map (\r -> map (read . pure :: Char -> Int) r) $
                 words $ fileContents
    let sols = map (\board -> reprBoard (fromJust (getSolution board)) ++ "\n") boards
    mapM_ putStr sols

File: js/solve.js
rows = Array.from(
    [...Array(9).keys()], i => Array.from(
        [...Array(9).keys()], j => 9*i+j
    )
)

cols = Array.from(
    [...Array(9).keys()], i => Array.from(
        [...Array(9).keys()], j => 9*j+i
    )
)

boxs = Array.from(
    [...Array(9).keys()], i => Array.from(
        [...Array(81).keys()].filter(j => 3*(j/27>>0) + (j%9/3)>>0==i)
    )
)

digits = Array.from(Array(9).keys()).map(x => x+1)

s = "000700800006000031040002000024070000010030080000060290000800070860000500002006000"
boardExample = Array.from(s).map(x => x*1)

function legalBoard(board) {
    let groups = [
        ...rows.map(r => r.map(x => board[x])),
        ...cols.map(c => c.map(x => board[x])),
        ...boxs.map(b => b.map(x => board[x])),
    ]
    for (var g of groups) {
        let digitCounts = Array(10).fill(0)
        for (var x of g) { digitCounts[x]++ }
//        console.log(g, digitCounts, Math.max(...digitCounts.slice(1,10)))
        if (Math.max(...digitCounts.slice(1,10)) > 1) return false
    }
    return true
}

function printBoard(board) {
    console.log(
        [...Array(9).keys()].map(
            x => board.slice(9*x, 9*x+9).map(String).join("")
        ).join("\n") + "\n"
    )
}

function solve(board) {
    if (board.indexOf(0) == -1) return board
    let boardCopy = Array.from(board)
    let currentPos = board.indexOf(0)
    for (let i=1; i<=9; i++) {
        board[currentPos] = i
        if (legalBoard(board)) {
            let temp = solve(board)
            if (temp.length > 0) return temp
        }
        board = Array.from(boardCopy)
    }
    return []
}

const fs = require('fs')
fs.readFile("../sudoku_puzzles", 'utf8' , (err, data) => {
    if (err) {
        console.error(err)
        return
    }
    let puzzles = data.split("\n").map(s => Array.from(s).map(Number))
    for (let puzzle of puzzles) {
        printBoard(solve(puzzle))
    }
})



